### PR TITLE
Create an Ideas & Suggestions page with voting  #582

### PR DESCRIPTION
**Pull request by Sarthak5598**
#582

### CODE REVIEW

**Comment by Sarthak5598** [comment]
Uploading BLT - Opera 2024-07-16 14-39-05.mp4…

**Comment by Sarthak5598** [comment]
https://github.com/user-attachments/assets/ea2ec51d-529a-438c-90ac-2cbece3c385e

File: blt/urls.py [hunk 1]
```diff
@@ -513,6 +517,10 @@
     path("fetch-current-bid/", fetch_current_bid, name="fetch_current_bid"),
     path("Submitpr/", submit_pr, name="submit_pr"),
     path("weekly-report/", weekly_report, name="weekly_report"),
+    path("add-suggestion/", add_suggestions, name="add_suggestions"),
+    path("view-suggestion/", view_suggestions, name="view_suggestions"),
+    path("vote-suggestion/", vote_suggestions, name="vote_suggestions"),
+    path("set-vote/", set_vote, name="set_vote"),
```
Comment (arkid15r) [inline blt/urls.py:523]: Could you group them under `/suggestion/` prefix?

File: website/admin.py [hunk 2]
```diff
@@ -22,6 +22,8 @@
     Points,
     Project,
     Subscription,
+    Suggestion,
+    SuggestionLikes,
```
Comment (arkid15r) [inline website/admin.py:26]: Be consistent -- use either likes or votes everywhere.

File: website/admin.py [hunk 3]
```diff
@@ -294,6 +296,14 @@ class ChatBotLogAdmin(admin.ModelAdmin):
     list_display = ("id", "question", "answer", "timestamp")
 
 
+class SuggestionsAdmin(admin.ModelAdmin):
```
Comment (arkid15r) [inline website/admin.py:299]:
```suggestion
class SuggestionAdmin(admin.ModelAdmin):
```

File: website/models.py [hunk 4]
```diff
@@ -638,6 +638,28 @@ def __str__(self):
         return f"Q: {self.question} | A: {self.answer} at {self.timestamp}"
 
 
+class Suggestion(models.Model):
+    user = models.CharField(max_length=100, default="Add user")
```
Comment (arkid15r) [inline website/models.py:642]: Should we use a Django user FK here?

File: website/models.py [hunk 5]
```diff
@@ -638,6 +638,28 @@ def __str__(self):
...
+    title = models.CharField(max_length=200)
+    description = models.TextField(max_length=1000, null=True, blank=True)
+    up_vote = models.IntegerField(null=True, blank=True, default=0)
```
Comment (arkid15r) [inline website/models.py:645]:
```suggestion
    up_votes = models.IntegerField(null=True, blank=True, default=0)
```
or `up_vote_count`

File: website/models.py [hunk 6]
```diff
@@ -638,6 +638,28 @@ def __str__(self):
...
+    down_vote = models.IntegerField(null=True, blank=True, default=0)
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
Comment (arkid15r) [inline website/models.py:647]: I'm not sure you need it. If you want uuid add it with a different name

File: website/models.py [hunk 7]
```diff
@@ -638,6 +638,28 @@ def __str__(self):
...
+
+    def __str__(self):
+        return f"{self.title} by {self.name}"
```
Comment (arkid15r) [inline website/models.py:650]: You don't have `name` attribute.

File: website/models.py [hunk 8]
```diff
@@ -638,6 +638,28 @@ def __str__(self):
...
+
+
+class SuggestionLikes(models.Model):
+    suggestion = models.ForeignKey(Suggestion, on_delete=models.CASCADE)
+    user = models.CharField(max_length=100, default="Add user")
```
Comment (arkid15r) [inline website/models.py:655]: Why not FK to user?

File: website/models.py [hunk 9]
```diff
@@ -638,6 +638,28 @@ def __str__(self):
...
+    up_vote = models.BooleanField(default=False)
+    down_vote = models.BooleanField(default=False)
```
Comment (arkid15r) [inline website/models.py:657]: optional: It looks excessive for just 2 mutually exclusive options.

File: website/templates/feature_suggestion.html [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+        const upvotesLabel = document.getElementById(`up_vote-${id}`);
+        const downvotesLabel = document.getElementById(`down_vote-${id}`);
+
+        if (!upvoteBtn || !downvoteBtn || !upvotesLabel || !downvotesLabel) {
+            console.error('Element not found for id:', id);
+            return;
+        }
+        
+        let up_vote;
+        let down_vote;
```
Comment (arkid15r) [inline website/templates/feature_suggestion.html:196]: JS uses camel case

File: website/views.py [hunk 11]
```diff
@@ -4744,3 +4746,92 @@ def blt_tomato(request):
...
+@login_required
+def vote_suggestions(request):
+    if request.method == "POST":
+        user = request.user.username
+        data = json.loads(request.body)
+        suggestion_id = data.get("suggestion_id")
+        suggestion = Suggestion.objects.get(id=suggestion_id)
+        up_vote = data.get("up_vote")
+        down_vote = data.get("down_vote")
+        liked = SuggestionLikes.objects.filter(user=user, suggestion=suggestion).exists()
```
Comment (arkid15r) [inline website/views.py:4760]: The name is misleading -- this one can be either upvote/downvote action.
```suggestion
        voted = SuggestionLikes.objects.filter(user=user, suggestion=suggestion).exists()
```

File: website/views.py [hunk 12]
```diff
@@ -4744,3 +4746,92 @@ def blt_tomato(request):
...
+                    )
+                    suggestion.down_vote += 1
+            elif down_vote is False:
+                suggestion.down_vote -= 1
+                liked = SuggestionLikes.objects.filter(user=user, suggestion=suggestion).delete()
+                if up_vote is True:
+                    liked = SuggestionLikes.objects.create(
+                        user=user, suggestion=suggestion, up_vote=True, down_vote=False
+                    )
+                    suggestion.up_vote += 1
```
Comment (arkid15r) [inline website/views.py:4788]: Could you simplify this using [update_or_create](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#update-or-create)?

File: website/views.py [hunk 13]
```diff
@@ -4744,3 +4746,92 @@ def blt_tomato(request):
...
+            "success": True,
+            "up_vote": suggestion.up_vote,
+            "down_vote": suggestion.down_vote,
+        }
+        return JsonResponse(response)
+
+    return JsonResponse({"success": False, "error": "Invalid request method"}, status=402)
+
+
+def set_vote(request):
```
Comment (arkid15r) [inline website/views.py:4800]: The name is misleading.

Should it be available for authenticated user only?

File: website/views.py [hunk 14]
```diff
@@ -4744,3 +4746,92 @@ def blt_tomato(request):
...
+    if request.method == "POST":
```
Comment (arkid15r) [inline website/views.py:4801]: Why not `GET`?

File: website/views.py [hunk 15]
```diff
@@ -4744,3 +4746,92 @@ def blt_tomato(request):
...
+        user = request.user.username
+        data = json.loads(request.body)
+        id = data.get("id")
+        suggestion = Suggestion.objects.get(id=id)
```
Comment (arkid15r) [inline website/views.py:4805]: What if suggestion doesn't exist?

**Review by arkid15r — DISMISSED** [review]
Please look into the following comments when you get a chance:

File: blt/urls.py (hunk 1 above)
Comment (Sarthak5598) [reply blt/urls.py:523]: Sure!

File: website/views.py (hunk 14 above)
Comment (Sarthak5598) [reply website/views.py:4801]: Isn;t it better to use post method?

File: website/models.py (hunk 6 above)
Comment (Sarthak5598) [reply website/models.py:647]: Isnt usig uuid a better option to create ids?

**Comment by Sarthak5598** [comment]
Thanks for the Review @arkid15r

File: website/views.py (hunk 13 above)
Comment (Sarthak5598) [reply website/views.py]: changed the name and also added `login_required` thanks!

File: website/views.py (hunk 12 above)
Comment (Sarthak5598) [reply website/views.py]: will look into it

File: website/models.py (hunk 9 above)
Comment (Sarthak5598) [reply website/models.py:657]: We can do it in one but as we need two fields in suggestions , I wanted to be consistent with it!

File: website/views.py (hunk 14 above)
Comment (arkid15r) [reply website/views.py:4802]: It depends. The `GET` should be fine here.

I suggest reading more about GET vs POST (maybe other methods too).

File: website/views.py (hunk 12 above)
Comment (arkid15r) [reply website/views.py]: Are you still looking into it? The code seems to be the same

File: website/models.py (hunk 6 above)
Comment (arkid15r) [reply website/models.py]: It depends. Both have their pros and cons.

**Review by arkid15r — COMMENTED** [review]
Thanks for updating the code!

File: website/admin.py [hunk 16]
```diff
@@ -327,6 +337,8 @@ class ProjectAdmin(admin.ModelAdmin):
 admin.site.register(IssueScreenshot, IssueScreenshotAdmin)
 admin.site.register(HuntPrize)
 admin.site.register(ChatBotLog, ChatBotLogAdmin)
+admin.site.register(Suggestion, SuggestionAdmin)
+admin.site.register(SuggestionVotes, SuggestionLikesAdmin)
```
Comment (arkid15r) [inline website/admin.py:341]:
```suggestion
admin.site.register(SuggestionVotes, SuggestionVotesAdmin)
```

File: website/views.py (hunk 12 above)
Comment (Sarthak5598) [reply website/views.py]: Yes , I am working on it !

**Review by DonnieBLT — DISMISSED** [review]
LGTM

File: website/views.py [hunk 17]
```diff
@@ -4749,3 +4751,108 @@ def blt_tomato(request):
...
+                    suggestion.down_votes += 1
+        else:
+            if up_vote is False:
+                suggestion.up_votes -= 1
+            if down_vote is False:
+                suggestion.down_votes -= 1
+
+            voted = SuggestionVotes.objects.filter(user=user, suggestion=suggestion).delete()
+
+            if up_vote is True:
```
Comment (arkid15r) [inline website/views.py:4787]:
```suggestion
            if up_vote:
```

File: website/views.py [hunk 18]
```diff
@@ -4749,3 +4751,108 @@ def blt_tomato(request):
...
+                voted = SuggestionVotes.objects.create(
+                    user=user, suggestion=suggestion, up_vote=True, down_vote=False
+                )
+                suggestion.up_votes += 1
+
+            if down_vote is True:
```
Comment (arkid15r) [inline website/views.py:4793]:
```suggestion
            if down_vote:
```

File: website/views.py [hunk 19]
```diff
@@ -4749,3 +4751,108 @@ def blt_tomato(request):
...
+                voted = SuggestionVotes.objects.create(
+                    user=user, suggestion=suggestion, up_vote=up_vote, down_vote=down_vote
+                )
+
+                if up_vote:
+                    suggestion.up_votes += 1
+                if down_vote:
+                    suggestion.down_votes += 1
+        else:
+            if up_vote is False:
```
Comment (arkid15r) [inline website/views.py:4780]:
```suggestion
            if not up_vote
```

**Review by DonnieBLT — APPROVED** [review]

**Comment by Sarthak5598** [comment]
yes everything is resolved @DonnieBLT

**Pull request merged**